File: global-templates/skills/testing/page-object-model/templates/base.page.ts
/**
 * Base Page Object class for Playwright tests (TypeScript).
 */
import { Page, Locator, expect } from '@playwright/test';

export abstract class BasePage {
  readonly page: Page;
  abstract readonly baseUrl: string;

  constructor(page: Page) {
    this.page = page;
  }

  async navigate(path: string = ''): Promise<void> {
    await this.page.goto(`${this.baseUrl}${path}`);
    await this.page.waitForLoadState('networkidle');
  }

  async getTitle(): Promise<string> {
    return this.page.title();
  }

  async getUrl(): Promise<string> {
    return this.page.url();
  }

  async takeScreenshot(name: string, fullPage: boolean = true): Promise<void> {
    await this.page.screenshot({
      path: `/tmp/${name}.png`,
      fullPage
    });
  }

  async waitForElement(selector: string, timeout: number = 5000): Promise<void> {
    await this.page.waitForSelector(selector, { timeout });
  }

  async waitForUrl(urlPattern: string | RegExp, timeout: number = 5000): Promise<void> {
    await this.page.waitForURL(urlPattern, { timeout });
  }

  async scrollToElement(selector: string): Promise<void> {
    await this.page.locator(selector).scrollIntoViewIfNeeded();
  }

  async isElementVisible(selector: string): Promise<boolean> {
    return this.page.locator(selector).isVisible();
  }

  async getText(selector: string): Promise<string> {
    return (await this.page.locator(selector).textContent()) ?? '';
  }

  async click(selector: string): Promise<void> {
    await this.page.locator(selector).click();
  }

  async fill(selector: string, value: string): Promise<void> {
    await this.page.locator(selector).fill(value);
  }

  async expectVisible(selector: string): Promise<void> {
    await expect(this.page.locator(selector)).toBeVisible();
  }

  async expectHidden(selector: string): Promise<void> {
    await expect(this.page.locator(selector)).toBeHidden();
  }

  async expectText(selector: string, text: string): Promise<void> {
    await expect(this.page.locator(selector)).toHaveText(text);
  }

  async expectUrl(path: string): Promise<void> {
    await expect(this.page).toHaveURL(`${this.baseUrl}${path}`);
  }
}

File: global-templates/skills/testing/page-object-model/templates/base_page.py
"""Base Page Object class for Playwright tests."""
from playwright.sync_api import Page, expect


class BasePage:
    """Base class for all Page Objects."""

    base_url: str = "http://localhost:3000"

    def __init__(self, page: Page):
        self.page = page

    def navigate(self, path: str = ""):
        """Navigate to a path relative to base_url."""
        self.page.goto(f"{self.base_url}{path}")
        self.page.wait_for_load_state("networkidle")
        return self

    def get_title(self) -> str:
        """Get page title."""
        return self.page.title()

    def get_url(self) -> str:
        """Get current URL."""
        return self.page.url

    def take_screenshot(self, name: str, full_page: bool = True):
        """Take a screenshot."""
        self.page.screenshot(path=f"/tmp/{name}.png", full_page=full_page)
        return self

    def wait_for_element(self, selector: str, timeout: int = 5000):
        """Wait for an element to be visible."""
        self.page.wait_for_selector(selector, timeout=timeout)
        return self

    def wait_for_url(self, url_pattern: str, timeout: int = 5000):
        """Wait for URL to match pattern."""
        self.page.wait_for_url(url_pattern, timeout=timeout)
        return self

    def scroll_to_element(self, selector: str):
        """Scroll element into view."""
        self.page.locator(selector).scroll_into_view_if_needed()
        return self

    def is_element_visible(self, selector: str) -> bool:
        """Check if element is visible."""
        return self.page.locator(selector).is_visible()

    def get_text(self, selector: str) -> str:
        """Get text content of an element."""
        return self.page.locator(selector).text_content() or ""
